File: docs/js/unit_mix.js
const RENT_ROLL_START_ROW = 8;
const RENT_ROLL_END_ROW = 67;

const UNIT_MIX_COLUMNS = [
  ["unit_type", "Unit Type (SF)", "text"],
  ["units", "# of Units", "integer"],
  ["occupied", "# of Occupied Units", "integer"],
  ["vacant", "# of Vacant Units", "integer"],
  ["pct", "%", "percent"],
  ["gpr", "GPR", "currency"],
  ["loss_to_lease", "Loss to Lease", "currency"],
  ["in_place", "In Place Rent", "currency"],
  ["vacancy", "Vacancy", "currency"],
  ["net", "Net Rental Income", "currency"],
];

function isVacantOrDown(name) {
  if (name == null) {
    return false;
  }
  const normalized = String(name).trim().toUpperCase();
  return normalized === "VACANT" || normalized === "DOWN";
}

function toFloat(value) {
  if (value == null || value === "") {
    return null;
  }
  const number = Number(value);
  return Number.isFinite(number) ? number : null;
}

function average(values) {
  if (!values.length) {
    return 0;
  }
  return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function cellValue(sheet, row, column) {
  const ref = XLSX.utils.encode_cell({ r: row - 1, c: column - 1 });
  const cell = sheet[ref];
  return cell ? cell.v : null;
}

function readRentRollRows(workbook) {
  const sheetName = workbook.SheetNames[0];
  const sheet = workbook.Sheets[sheetName];
  const rows = [];

  for (let rowIdx = RENT_ROLL_START_ROW; rowIdx <= RENT_ROLL_END_ROW; rowIdx += 1) {
    const sqFt = cellValue(sheet, rowIdx, 3);
    if (sqFt == null || String(sqFt).trim() === "") {
      continue;
    }

    rows.push({
      sq_ft: String(sqFt).trim(),
      name: cellValue(sheet, rowIdx, 5),
      market_rent: cellValue(sheet, rowIdx, 6),
      actual_rent: cellValue(sheet, rowIdx, 7),
    });
  }

  return rows;
}

function uniqueUnitTypes(rows) {
  const seen = new Map();
  for (const row of rows) {
    if (!seen.has(row.sq_ft)) {
      seen.set(row.sq_ft, true);
    }
  }
  return Array.from(seen.keys());
}

function calculateUnitMixFromWorkbook(workbook) {
  const rentRows = readRentRollRows(workbook);
  if (!rentRows.length) {
    throw new Error("No rent roll data found in rows 8-67 (column C).");
  }

  const unitTypes = uniqueUnitTypes(rentRows);
  const dataRows = unitTypes.map((unitType) => {
    const subset = rentRows.filter((row) => row.sq_ft === unitType);
    const units = subset.length;
    const vacant = subset.filter((row) => isVacantOrDown(row.name)).length;
    const occupied = units - vacant;

    const marketRents = subset
      .map((row) => toFloat(row.market_rent))
      .filter((value) => value != null);
    const gpr = average(marketRents);

    const occupiedRows = subset.filter((row) => !isVacantOrDown(row.name));
    const inPlaceValues = occupiedRows
      .map((row) => toFloat(row.actual_rent))
      .filter((value) => value != null);
    const inPlace = average(inPlaceValues);

    const actualValues = subset
      .map((row) => toFloat(row.actual_rent))
      .filter((value) => value != null);
    const net = average(actualValues);

    return {
      unit_type: unitType,
      units,
      occupied,
      vacant,
      pct: 0,
      gpr,
      loss_to_lease: inPlace - gpr,
      in_place: inPlace,
      vacancy: net - inPlace,
      net,
    };
  });

  const totalUnits = dataRows.reduce((sum, row) => sum + row.units, 0);
  for (const row of dataRows) {
    row.pct = totalUnits ? row.units / totalUnits : 0;
  }

  const totalOccupied = dataRows.reduce((sum, row) => sum + row.occupied, 0);
  const totalVacant = dataRows.reduce((sum, row) => sum + row.vacant, 0);

  const monthlyGpr = dataRows.reduce((sum, row) => sum + row.gpr * row.units, 0);
  const monthlyLoss = dataRows.reduce(
    (sum, row) => sum + row.loss_to_lease * row.units,
    0
  );
  const monthlyInPlace = dataRows.reduce(
    (sum, row) => sum + row.in_place * row.units,
    0
  );
  const monthlyNet = dataRows.reduce((sum, row) => sum + row.net * row.units, 0);
  const monthlyVacancy = monthlyNet - monthlyInPlace;

  return {
    title: "Unit Mix",
    columns: UNIT_MIX_COLUMNS.map(([key, label, format]) => ({
      key,
      label,
      format,
    })),
    rows: dataRows,
    totals: {
      label: "Totals",
      units: totalUnits,
      occupied: totalOccupied,
      vacant: totalVacant,
    },
    monthly: {
      label: "Monthly",
      gpr: monthlyGpr,
      loss_to_lease: monthlyLoss,
      in_place: monthlyInPlace,
      vacancy: monthlyVacancy,
      net: monthlyNet,
    },
    annually: {
      label: "Annually",
      gpr: monthlyGpr * 12,
      loss_to_lease: monthlyLoss * 12,
      in_place: monthlyInPlace * 12,
      vacancy: monthlyVacancy * 12,
      net: monthlyNet * 12,
    },
  };
}

async function calculateUnitMixFromFile(file) {
  const buffer = await file.arrayBuffer();
  const workbook = XLSX.read(buffer, { type: "array" });
  return calculateUnitMixFromWorkbook(workbook);
}

File: unit_mix.py
"""Unit Mix calculator (Proforma E1:N28) from rent roll Excel files."""

from __future__ import annotations

from collections import OrderedDict
from io import BytesIO
from typing import Any

from openpyxl import load_workbook

RENT_ROLL_START_ROW = 8
RENT_ROLL_END_ROW = 67

COLUMNS = [
    ("unit_type", "Unit Type (SF)", "text"),
    ("units", "# of Units", "integer"),
    ("occupied", "# of Occupied Units", "integer"),
    ("vacant", "# of Vacant Units", "integer"),
    ("pct", "%", "percent"),
    ("gpr", "GPR", "currency"),
    ("loss_to_lease", "Loss to Lease", "currency"),
    ("in_place", "In Place Rent", "currency"),
    ("vacancy", "Vacancy", "currency"),
    ("net", "Net Rental Income", "currency"),
]


def _is_vacant_or_down(name: Any) -> bool:
    if name is None:
        return False
    return str(name).strip().upper() in ("VACANT", "DOWN")


def _to_float(value: Any) -> float | None:
    if value is None or value == "":
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _average(values: list[float]) -> float:
    return sum(values) / len(values) if values else 0.0


def _read_rent_roll_rows(file_bytes: bytes) -> list[dict[str, Any]]:
    workbook = load_workbook(BytesIO(file_bytes), data_only=True, read_only=True)
    sheet = workbook.active

    rows: list[dict[str, Any]] = []
    for row_idx in range(RENT_ROLL_START_ROW, RENT_ROLL_END_ROW + 1):
        sq_ft = sheet.cell(row=row_idx, column=3).value  # column C
        if sq_ft is None or str(sq_ft).strip() == "":
            continue

        rows.append(
            {
                "sq_ft": str(sq_ft).strip(),
                "name": sheet.cell(row=row_idx, column=5).value,  # column E
                "market_rent": sheet.cell(row=row_idx, column=6).value,  # column F
                "actual_rent": sheet.cell(row=row_idx, column=7).value,  # column G
            }
        )

    workbook.close()
    return rows


def _unique_unit_types(rows: list[dict[str, Any]]) -> list[str]:
    seen: OrderedDict[str, None] = OrderedDict()
    for row in rows:
        seen.setdefault(row["sq_ft"], None)
    return list(seen.keys())


def calculate_unit_mix(file_bytes: bytes) -> dict[str, Any]:
    rent_rows = _read_rent_roll_rows(file_bytes)
    if not rent_rows:
        raise ValueError("No rent roll data found in rows 8–67 (column C).")

    unit_types = _unique_unit_types(rent_rows)
    data_rows: list[dict[str, float | str]] = []

    for unit_type in unit_types:
        subset = [row for row in rent_rows if row["sq_ft"] == unit_type]
        units = len(subset)
        vacant = sum(1 for row in subset if _is_vacant_or_down(row["name"]))
        occupied = units - vacant

        market_rents = [
            value
            for row in subset
            if (value := _to_float(row["market_rent"])) is not None
        ]
        gpr = _average(market_rents)

        occupied_rows = [
            row for row in subset if not _is_vacant_or_down(row["name"])
        ]
        in_place_values = [
            value
            for row in occupied_rows
            if (value := _to_float(row["actual_rent"])) is not None
        ]
        in_place = _average(in_place_values)

        actual_values = [
            value
            for row in subset
            if (value := _to_float(row["actual_rent"])) is not None
        ]
        net = _average(actual_values)

        data_rows.append(
            {
                "unit_type": unit_type,
                "units": units,
                "occupied": occupied,
                "vacant": vacant,
                "pct": 0.0,
                "gpr": gpr,
                "loss_to_lease": in_place - gpr,
                "in_place": in_place,
                "vacancy": net - in_place,
                "net": net,
            }
        )

    total_units = sum(int(row["units"]) for row in data_rows)
    for row in data_rows:
        row["pct"] = row["units"] / total_units if total_units else 0.0

    total_occupied = sum(int(row["occupied"]) for row in data_rows)
    total_vacant = sum(int(row["vacant"]) for row in data_rows)

    monthly_gpr = sum(row["gpr"] * row["units"] for row in data_rows)
    monthly_loss = sum(row["loss_to_lease"] * row["units"] for row in data_rows)
    monthly_in_place = sum(row["in_place"] * row["units"] for row in data_rows)
    monthly_net = sum(row["net"] * row["units"] for row in data_rows)
    monthly_vacancy = monthly_net - monthly_in_place

    return {
        "title": "Unit Mix",
        "columns": [{"key": key, "label": label, "format": fmt} for key, label, fmt in COLUMNS],
        "rows": data_rows,
        "totals": {
            "label": "Totals",
            "units": total_units,
            "occupied": total_occupied,
            "vacant": total_vacant,
        },
        "monthly": {
            "label": "Monthly",
            "gpr": monthly_gpr,
            "loss_to_lease": monthly_loss,
            "in_place": monthly_in_place,
            "vacancy": monthly_vacancy,
            "net": monthly_net,
        },
        "annually": {
            "label": "Annually",
            "gpr": monthly_gpr * 12,
            "loss_to_lease": monthly_loss * 12,
            "in_place": monthly_in_place * 12,
            "vacancy": monthly_vacancy * 12,
            "net": monthly_net * 12,
        },
    }
